pagerduty/resource_pagerduty_tag_assignment.go

```diff
@@ -0,0 +1,189 @@
+		// they receive no tag id from the PagerDuty API at this time
+		Type:       "tag_reference",
+		EntityType: d.Get("entity_type").(string),
+		EntityID:   d.Get("entity_id").(string),
+	}
+	if attr, ok := d.GetOk("tag_id"); ok {
+		assignment.TagID = attr.(string)
+	}
+
+	return assignment
+}
+
+func resourcePagerDutyTagAssignmentCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	assignment := buildTagAssignmentStruct(d)
+	assignments := &pagerduty.TagAssignments{
+		Add: []*pagerduty.TagAssignment{assignment},
+	}
+
+	log.Printf("[INFO] Creating PagerDuty tag assignment with tagID %s for %s entity with ID %s", assignment.TagID, assignment.EntityType, assignment.EntityID)
+
+	retryErr := resource.Retry(10*time.Second, func() *resource.RetryError {
+		if _, err := client.Tags.Assign(assignment.EntityType, assignment.EntityID, assignments); err != nil {
+			if isErrCode(err, 400) || isErrCode(err, 429) {
+				return resource.RetryableError(err)
+			}
+
+			return resource.NonRetryableError(err)
+		} else {
+			// create tag_assignment id using the entityID.tagID as PagerDuty API does not return one
+			assignmentID := fmt.Sprintf("%v.%v", assignment.EntityID, assignment.TagID)
+			d.SetId(assignmentID)
+		}
+		return nil
+	})
+
+	if retryErr != nil {
+		return retryErr
+	}
+	// give PagerDuty 2 seconds to save the assignment correctly
+	time.Sleep(2 * time.Second)
+	return resourcePagerDutyTagAssignmentRead(d, meta)
+
+}
+
+func resourcePagerDutyTagAssignmentRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	assignment := buildTagAssignmentStruct(d)
+
+	log.Printf("[INFO] Reading PagerDuty tag assignment with tagID %s for %s entity with ID %s", assignment.TagID, assignment.EntityType, assignment.EntityID)
+
+	return resource.Retry(30*time.Second, func() *resource.RetryError {
+		if tagResponse, _, err := client.Tags.ListTagsForEntity(assignment.EntityType, assignment.EntityID); err != nil {
+			time.Sleep(2 * time.Second)
+			return resource.RetryableError(err)
+		} else if tagResponse != nil {
+			var foundTag *pagerduty.Tag
+
+			// loop tags and find matching ID
+			for _, tag := range tagResponse.Tags {
+				if tag.ID == assignment.TagID {
+					foundTag = tag
+					break
+				}
+			}
+			if foundTag == nil {
+				d.SetId("")
+				return nil
+			}
+		}
+		return nil
+	})
+}
+
+func resourcePagerDutyTagAssignmentDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	assignment := buildTagAssignmentStruct(d)
+	assignments := &pagerduty.TagAssignments{
+		Remove: []*pagerduty.TagAssignment{assignment},
+	}
+	log.Printf("[INFO] Deleting PagerDuty tag assignment with tagID %s for entityID %s", assignment.TagID, assignment.EntityID)
+
+	retryErr := resource.Retry(10*time.Second, func() *resource.RetryError {
```

Comment:
We could also return the result of `Retry()` here I think like we do in the `Read()` func 🙂 
```go
	return resource.Retry(10*time.Second, func() *resource.RetryError {
		if _, err := client.Tags.Assign(assignment.EntityType, assignment.EntityID, assignments); err != nil {
			if isErrCode(err, 400) || isErrCode(err, 429) {
				return resource.RetryableError(err)
			}

			return resource.NonRetryableError(err)
		} else {
			d.SetId("")
		}
		return nil
	})

```